File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.py
def solution(friends, gifts):
    answer = 0
    n=len(friends)
    dic={}
    for i in range(n):
        dic[friends[i]]=i
    present=[[0 for _ in range(n)] for _ in range(n)]
    for i in gifts:
        a,b=i.split(" ")
        present[dic[a]][dic[b]]+=1
    pp=[]
    for i in range(n):
        cnt=0
        for j in range(n):
            cnt+=present[i][j]
            cnt-=present[j][i]
        pp.append(cnt)
    total=[0]*n
    for i in range(n):
        for j in range(n):
            if j>i:
                if present[i][j]<present[j][i]:
                    total[j]+=1
                elif present[i][j]==present[j][i]:
                    if pp[i]>pp[j]:
                        total[i]+=1
                    elif pp[j]>pp[i]:
                        total[j]+=1
                else:
                    total[i]+=1
    return max(total)

File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.java
import java.util.HashMap;
class Solution {
    public int solution(String[] friends, String[] gifts) {
        int answer = 0;
        int friendsLength=friends.length;
        HashMap<String, Integer> dic=new HashMap<>();
        int[] giftDegree=new int[friendsLength];
        int[][] giftGraph= new int[friendsLength][friendsLength];
        for (int i=0;i<friendsLength;i++) {
            dic.put(friends[i],i);
        }
        for (String gift : gifts) {
            String[] split=gift.split(" ");
            String x=split[0];
            String y=split[1];
            giftDegree[dic.get(x)]++;
            giftDegree[dic.get(y)]--;
            giftGraph[dic.get(x)][dic.get(y)]++;
        }
        for (int i=0;i<friendsLength;i++) {
            int n=0;
            for (int j=0; j<friendsLength; j++) {
                if (i==j) {
                    continue;
                }
                if (giftGraph[i][j]>giftGraph[j][i] || (giftGraph[i][j]==giftGraph[j][i] && giftDegree[i]>giftDegree[j])) {
                    n++;
                }
            }
            if (answer<n) {
                answer=n;
            }
        }
        return answer;
    }
}